File: backend/observability.py
"""
Sentry initialization for the backend service.

Reads:
  SENTRY_DSN            — DSN URL. If unset/empty, init_sentry() is a no-op.
  SENTRY_ENVIRONMENT    — e.g. "production", "staging", "local". Default: "local".
  SENTRY_RELEASE        — git SHA injected at container build (see Dockerfile).
  SENTRY_TRACES_SAMPLE_RATE — float 0.0-1.0. Default 0.05 (5%).

Behavior:
  - No DSN → no-op. Safe to call from any service. No network, no overhead.
  - With DSN → captures unhandled exceptions, FastAPI request errors, asyncpg
    pool errors. Filters known noise (healthchecks, expected 4xx).
  - Never sends request bodies or headers that could contain PII.
"""
import logging
import os
from typing import Any, Dict, Optional

logger = logging.getLogger(__name__)

# Endpoints / status codes we don't want to alert on.
_NOISY_PATHS = {"/health", "/healthz", "/metrics", "/readyz"}
_EXPECTED_STATUS = {401, 403, 404, 429}


def _before_send(event: Dict[str, Any], hint: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """Drop noise before it ever leaves the process."""
    # Drop healthcheck / metrics request errors.
    request = event.get("request") or {}
    url = request.get("url") or ""
    for noisy in _NOISY_PATHS:
        if url.endswith(noisy):
            return None

    # Drop expected HTTP status errors (4xx that the app handles).
    exc_info = hint.get("exc_info")
    if exc_info:
        exc = exc_info[1]
        status_code = getattr(exc, "status_code", None)
        if status_code in _EXPECTED_STATUS:
            return None

    # Strip cookies and Authorization header just in case.
    headers = request.get("headers") or {}
    if isinstance(headers, dict):
        headers.pop("cookie", None)
        headers.pop("Cookie", None)
        headers.pop("authorization", None)
        headers.pop("Authorization", None)
    return event


def init_sentry(service_name: str = "backend") -> bool:
    """
    Initialize Sentry if SENTRY_DSN is set. Returns True if initialized.

    Idempotent: safe to call multiple times. Returns False on no-op (no DSN).
    """
    dsn = os.environ.get("SENTRY_DSN", "").strip()
    if not dsn:
        logger.info("sentry: SENTRY_DSN unset, skipping init")
        return False

    try:
        import sentry_sdk
        from sentry_sdk.integrations.asyncio import AsyncioIntegration
        from sentry_sdk.integrations.fastapi import FastApiIntegration
        from sentry_sdk.integrations.starlette import StarletteIntegration
    except ImportError:
        logger.warning("sentry: sentry-sdk not installed, skipping init")
        return False

    environment = os.environ.get("SENTRY_ENVIRONMENT", "local")
    release = os.environ.get("SENTRY_RELEASE") or None
    try:
        traces_sample_rate = float(os.environ.get("SENTRY_TRACES_SAMPLE_RATE", "0.05"))
    except ValueError:
        traces_sample_rate = 0.05

    sentry_sdk.init(
        dsn=dsn,
        environment=environment,
        release=release,
        traces_sample_rate=traces_sample_rate,
        # Never auto-capture request bodies — they may contain email addresses
        # or rule patterns that count as customer data.
        send_default_pii=False,
        max_request_body_size="never",
        attach_stacktrace=True,
        integrations=[
            StarletteIntegration(transaction_style="endpoint"),
            FastApiIntegration(transaction_style="endpoint"),
            AsyncioIntegration(),
        ],
        before_send=_before_send,
    )
    sentry_sdk.set_tag("service", service_name)
    logger.info(
        "sentry: initialized service=%s environment=%s release=%s traces=%s",
        service_name, environment, release or "(unset)", traces_sample_rate,
    )
    return True

File: dashboard/src/observability.js
/**
 * Sentry initialization for the dashboard (React SPA).
 *
 * Reads from Vite env (must be prefixed VITE_ to be exposed to the bundle):
 *   VITE_SENTRY_DSN              — DSN URL. Unset/empty → no-op.
 *   VITE_SENTRY_ENVIRONMENT      — "production" | "staging" | "local". Default: "local".
 *   VITE_SENTRY_RELEASE          — git SHA injected at build time.
 *   VITE_SENTRY_TRACES_SAMPLE_RATE — float 0.0-1.0. Default 0.05.
 *
 * Privacy:
 *   - No PII attached automatically (Sentry's `sendDefaultPii: false`).
 *   - Network breadcrumbs strip Authorization headers via beforeBreadcrumb.
 *   - Avoid Replay for now — it can capture form inputs (auth, rule patterns).
 */
import * as Sentry from '@sentry/react';

const NOISY_URLS = ['/health', '/healthz', '/metrics'];
const EXPECTED_STATUS = new Set([401, 403, 404, 429]);

function beforeSend(event, hint) {
  // Drop noisy URLs.
  const url = event?.request?.url || '';
  if (NOISY_URLS.some((n) => url.endsWith(n))) return null;

  // Drop expected HTTP status errors thrown by axios/fetch wrappers.
  const status =
    hint?.originalException?.response?.status ??
    hint?.originalException?.status;
  if (status && EXPECTED_STATUS.has(status)) return null;

  // Strip Authorization header from request snapshot if present.
  if (event?.request?.headers) {
    delete event.request.headers.authorization;
    delete event.request.headers.Authorization;
    delete event.request.headers.cookie;
    delete event.request.headers.Cookie;
  }
  return event;
}

function beforeBreadcrumb(breadcrumb) {
  // Strip auth headers from fetch/xhr breadcrumbs.
  if (breadcrumb?.data?.request_headers) {
    delete breadcrumb.data.request_headers.authorization;
    delete breadcrumb.data.request_headers.Authorization;
  }
  return breadcrumb;
}

export function initSentry() {
  const dsn = import.meta.env.VITE_SENTRY_DSN;
  if (!dsn) {
    // eslint-disable-next-line no-console
    console.info('sentry: VITE_SENTRY_DSN unset, skipping init');
    return false;
  }

  const environment = import.meta.env.VITE_SENTRY_ENVIRONMENT || 'local';
  const release = import.meta.env.VITE_SENTRY_RELEASE || undefined;
  const tracesSampleRate = parseFloat(
    import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE || '0.05'
  );

  Sentry.init({
    dsn,
    environment,
    release,
    tracesSampleRate: Number.isFinite(tracesSampleRate) ? tracesSampleRate : 0.05,
    sendDefaultPii: false,
    integrations: [Sentry.browserTracingIntegration()],
    beforeSend,
    beforeBreadcrumb,
  });
  Sentry.setTag('service', 'dashboard');
  return true;
}
